content/courses/Ada_For_The_C_Embedded_Developer/chapters/08_Performance.rst

```diff
@@ -29,10 +29,103 @@ suspects and their resolution.
 Switches and optimizations
 --------------------------
 
-.. todo::
-
-    Complete section!
-
+Clever use of compilation switches might optimize the performance of an
+application significantly. In this section, we'll briefly look into some of
+the switches available in the GNAT toolchain.
+
+Optimizations levels
+~~~~~~~~~~~~~~~~~~~~
+
+Optimization levels can be found in many compilers for multiple languages. On
+the lowest level, the compiler doesn't optimize the code at all, while at the
+highest level, the compiler analyses the code and optimizes it by removing
+unnecessary operations and making most use of the target processor's
+capabilities.
+
+By being part of GCC, GNAT offers the same ``-O_`` switches as GCC:
+
++-------------+--------------------------------------------------------------+
+| Switch      | Description                                                  |
++=============+==============================================================+
+| ``-O0``     | No optimization: the generated code is completely            |
+|             | unoptimized. This is the default optimization level.         |
++-------------+--------------------------------------------------------------+
+| ``-O1``     | Moderate optimization.                                       |
++-------------+--------------------------------------------------------------+
+| ``-O2``     | Full optimization.                                           |
++-------------+--------------------------------------------------------------+
+| ``-O3``     | Same optimization level as for ``-O2``. In addition, further |
+|             | optimization strategies, such as aggressive automatic        |
+|             | inlining and vectorization.                                  |
++-------------+--------------------------------------------------------------+
+
+Note that the highest the level, the slowest will be the compilation time. For
+fast compilation during development phase, unless you're working on
+benchmarking algorithms, using ``-O0`` is probably a good idea.
+
+In addition to the levels presented above, GNAT also has the ``-Os`` switch,
+which allows for optimizing code and data usage.
+
+Inlining
+~~~~~~~~
+
+As we've seen in the previous section, the highest optimization level (``-O3``)
+tries to automatic inline subprograms whenever possible. There might be
+situations, however, where this doesn't happen automatically. In those cases,
+manual inlining may help.
```

Comment:
The problem with O3 is that it inlines too much rather than not enough.  As a result, the cache becomes an issue and the code can slow down as a result.